front/src/modules/simulationResult/components/SpaceTimeChart/useLazyProjectTrains.ts

```
@@ -0,0 +1,158 @@
+
+  const trainSchedulesById = useMemo(() => keyBy(trainSchedules, 'id'), [trainSchedules]);
+
+  useEffect(() => {
+    // reset the state when all the trains have been projected
+    if (
+      !moreTrainsToCome &&
+      readyToProjectNextPackage &&
+      nextPackageToProjectIndex * 10 > trainIdsToProject.length
+    ) {
+      setNextPackageToProjectIndex(0);
+      setTrainIdsToProject([]);
+    }
+  }, [moreTrainsToCome, readyToProjectNextPackage]);
+
+  // gradually project the trains on the selected path
+  useEffect(() => {
+    const projectNextPackage = async (
+      _infraId: number,
+      _pathfinding: PathfindingResultSuccess,
+      _trainIdsToProject: number[],
+      _nextPackageToProjectIndex: number
+    ) => {
+      setReadyToProjectNextPackage(false);
+
+      try {
+        const packageToProject = getTrainIdsPackage(
+          _nextPackageToProjectIndex,
+          _trainIdsToProject,
+          BATCH_SIZE
+        );
+
+        if (packageToProject.length === 0) {
+          setReadyToProjectNextPackage(true);
+          return;
+        }
+        setNextPackageToProjectIndex((prev) => prev + 1);
+
+        const { blocks, routes, track_section_ranges } = _pathfinding;
+        const rawProjectedTrains = await postV2TrainScheduleProjectPath({
+          projectPathForm: {
+            infra_id: _infraId,
+            ids: packageToProject,
+            path: { blocks, routes, track_section_ranges },
+            electrical_profile_set_id: electricalProfileSetId,
+          },
+        }).unwrap();
+
+        setProjectedTrainsById((prevTrains) => {
+          const newProjectedTrains = upsertNewProjectedTrains(
+            prevTrains,
+            rawProjectedTrains,
+            trainSchedulesById
+          );
+          return newProjectedTrains;
+        });
+        // console.log('2. package', packageToProject);
+      } catch (e) {
+        console.error('error', e);
```

Should we bubble up this error to the user, since it's coming from the API?

---

I don't really know how to do this properly right now, I would prefer to leave this for later, when we will refactor the notification system, if it's fine for you